README.rst
===========
MemSQL Collectd Plugin
===========

This libary implements a MemSQL Ops specific plugin for `collectd`.

Install
=======

.. code:: bash

    pip install memsql-collectd

Then add the following plugin to your collectd configuration[1]_:

.. [1] The collectd config file may be found at /etc/collectd.conf or /etc/collectd/collectd.conf if collectd was installed through a package manager.

.. code:: xml

    <LoadPlugin python>
        Globals true
    </LoadPlugin>

    <Plugin python>
        Import "memsql_collectd.plugin"
        <Module "memsql_collectd.plugin">
            TypesDB "/usr/share/collectd/types.db"
            Host "MASTER AGGREGATOR HOSTNAME/IP_ADDRESS"
            Port "3306"
            MemSQLNode True
        </Module>
    </Plugin>


CHANGES.txt
2014-12-11 Version 0.2.6

    * Fixed the hash function which produces classifier_ids so it matches up exactly with the one before 0.2.4

2014-10-27 Version 0.2.5

    * Added the option to disable the usage of du to scan MemSQL disk usage

2014-07-30 Version 0.2.4

    * Now supports .'s showing up inside part of a classifier (ie. a . can be inside plugin or
      plugin_instance)

2014-05-12 Version 0.2.3

    * MemSQL specific du/df extensions
        * If the node is a MemSQL Node we record the sizes of specific directories and files
          for a better ops experience.
    * Cleaned up the exit logic, and improved logging

2014-05-12 Version 0.2.2

    * Added classifier whitelist functionality

2014-04-13 Version 0.2.1

    * Fixed a bug that caused facts to never update

2014-03-26 Version 0.2.0

    * Switching to timezone invariant support for Ops 2.10.0 and above
    * Allow user to help the plugin find the correct MemSQL node

2014-02-20 Version 0.1.3

    * Fixed a bug in the multi inserts

2014-01-03 Version 0.1.2

    * Upgrade to memsql-python 2.9.0

2013-12-10 Version 0.1.1

    * Track more memsql values as counters
    * Allow override for node detection
    * Print version number at startup

2013-12-04 Version 0.1.0

    * Added config which disables counter/derive derivatives.
    * Added classifier blacklist functionality
    * Blacklist certain tmpfs's by default
        * /dev/shm, /dev, /run, /run/lock, /run/shm, /proc

2013-11-01 Version 0.0.2

    * Implemented prefix support
        * This will prefix all metrics sent with the specified string

2013-10-29 Version 0.0.1

    * Initial release.


memsql_collectd/cluster.py
from netifaces import interfaces, ifaddresses, AF_INET
import re
import socket
from memsql.common.connection_pool import ConnectionPool, PoolConnectionException
from memsql.common import database

# Status variables specified in the following array will be sent to
# collectd as COUNTERS as well as GAUGES.
# This will cause them to be stored as both an absolute value and a
# derivative.
COUNTER_STATUS_VARIABLES = [
    "Rows_affected_by_writes",
    "Rows_returned_by_reads",
    "Successful_write_queries",
    "Successful_read_queries",
    "Failed_read_queries",
    "Failed_write_queries",
    "Execution_time_of_reads",
    "Execution_time_of_write",
    "Row_lock_wait_time"
]

def find_node(connection_pool, possible_hostname=None, memsqlnode=None):
    if isinstance(memsqlnode, basestring):
        return find_node_by_name(connection_pool, memsqlnode)
    else:
        return find_node_by_address(connection_pool, possible_hostname)

def find_node_by_name(connection_pool, memsqlnode):
    """ Look up a node using the host:port specifier in memsqlnode """
    host, port = memsqlnode.split(':') if ':' in memsqlnode else (memsqlnode, '%')

    try:
        with connection_pool.connect() as conn:
            node_row = conn.get('''
                SELECT id, host, port
                FROM nodes
                WHERE nodes.host = %s AND nodes.port LIKE %s
                ORDER BY host, port LIMIT 1
            ''', host, port)
    except (PoolConnectionException, database.DatabaseError):
        return None

    if node_row is not None:
        return Node(node_row)

def find_node_by_address(connection_pool, possible_hostname=None):
    addresses = _network_addresses()

    if possible_hostname is not None:
        addresses.append(socket.gethostbyname(possible_hostname))

    # Select the matching host from the Dashboard node table that match any
    # of our ips. Hopefully this resolves to a single node.
    try:
        with connection_pool.connect() as conn:
            node_row = conn.get('''
                SELECT id, host, port
                FROM nodes
                WHERE nodes.host IN (%s)
                ORDER BY host, port LIMIT 1
            ''' % ','.join("'%s'" % addr for addr in addresses))
    except (PoolConnectionException, database.DatabaseError):
        return None

    if node_row is None:
        # we may be the master node, and the dashboard may be pointing at 127.0.0.1
        with connection_pool.connect_master() as conn:
            master_host = socket.gethostbyname(conn.connection_info()[0])
            if master_host in addresses:
                # ok we are the master
                node_row = conn.get('''
                    SELECT id, port, %s AS host
                    FROM nodes WHERE master=1
                ''', master_host)

    if node_row is not None:
        return Node(node_row)

def _network_addresses():
    ret = []
    for interface in interfaces():
        details = ifaddresses(interface)
        if AF_INET in details:
            for link in details[AF_INET]:
                if link['addr'] != '127.0.0.1':
                    ret.append(link['addr'])
    return ret

class Node(object):
    def __init__(self, node_row):
        self.update_from_node(node_row)
        self.alias = None
        self._pool = ConnectionPool()

    def update_from_node(self, node):
        self.id = node.id
        self.host = node.host
        self.port = node.port

    def update_alias(self, connection_pool, alias):
        self.alias = alias

        try:
            conn = connection_pool.connect_master()

            if conn:
                conn.execute('''
                    INSERT INTO node_alias (node_id, alias)
                    VALUES (%s, %s)
                    ON DUPLICATE KEY UPDATE alias=VALUES(alias)
                ''', self.id, alias)
        finally:
            if conn:
                conn.close()

    def connect(self):
        return self._pool.connect(
            host=self.host,
            port=self.port,
            user="dashboard",
            password="",
            database="information_schema")

    def status(self):
        with self.connect() as conn:
            rows = conn.query('SHOW STATUS EXTENDED')

        for row in rows:
            name = row.Variable_name
            try:
                value = self._parse_value(row.Value)
            except ValueError:
                continue
            yield (name, value)

    STATUS_CONSTS = re.compile(r"ms|MB|KB", re.I)

    def _parse_value(self, value):
        if self.STATUS_CONSTS.search(value):
            return float(value.split(" ")[0])
        return float(value)

    def variables(self):
        with self.connect() as conn:
            rows = conn.query('SHOW VARIABLES')
        for row in rows:
            yield (row.Variable_name, row.Value)
        for name, path in self.directories().iteritems():
            yield (name, path)

    def directories(self):
        with self.connect() as conn:
            rows = conn.query("SHOW STATUS EXTENDED LIKE '%_directory'")
        return dict([(row.Variable_name, row.Value) for row in rows])


memsql_collectd/analytics.py
from datetime import datetime
import threading
import math
from wraptor.decorators import throttle
import calendar
from itertools import repeat, chain, islice, tee, imap, ifilter
from memsql_collectd import util

CLASSIFIER_MIN_HASH_LEN = 6
CLASSIFIERS = [ 'instance_id', 'alpha', 'beta', 'gamma', 'delta', 'epsilon', 'zeta', 'eta', 'theta', 'iota', 'kappa', 'lambda', 'mu', 'nu', 'xi', 'omicron' ]
CLASSIFIER_COLUMNS = CLASSIFIERS + [ "classifier", "id", "last_updated" ]

ANALYTICS_COLUMNS = [ "classifier_id", "created", "value" ]
ANALYTICS_TEMPLATE = ["(%s, %s, %s)"]

def partition(iterator, cb):
    """ partitions iterator into two lists by running callback on each value
        * callback is only executed once per value
        * iterator must be finite, as this is a non-lazy partition function
    """
    trues, falses = tee(imap(lambda x: (cb(x), x), iterator), 2)
    return [v for r, v in trues if r], [v for r, v in falses if not r]

class AnalyticsRow(object):
    def __init__(self, created, raw_value, *classifier):
        self.created = created
        self.raw_value = raw_value

        self.value = None
        self._classifier_id = None

        self.classifier = tuple(islice(chain(ifilter(None, classifier), repeat('')), len(CLASSIFIERS)))

    @property
    def classifier_id(self):
        if self._classifier_id is None:
            parts = [x for x in self.classifier if x != '']
            if len(parts) < CLASSIFIER_MIN_HASH_LEN:
                diff = CLASSIFIER_MIN_HASH_LEN - len(parts)
                parts = parts + ([''] * diff)
            self._classifier_id = util.hash_64_bit(*parts)
        return self._classifier_id

    @property
    def joined_classifier(self):
        return '.'.join([x for x in self.classifier if x != ''])

    def classifier_values(self):
        """ Returns the entire row as a tuple for the classifiers table. """
        return self.classifier + (self.joined_classifier, self.classifier_id)

    def analytics_values(self):
        """ Returns a tuple for inserting into the analytics table. """
        return (self.classifier_id, int(self.created), self.value)

    def valid(self):
        """ Validates that the value of this row is a valid value (ie. not None and not NaN) """
        return not (self.value is None or math.isnan(self.value))

class AnalyticsCache(object):
    """ A thread safe cache for analytics rows.  Can efficiently flush
    all values in the cache using a multi-insert.
    """
    def __init__(self):
        self._lock = threading.RLock()
        self._pending = []
        self._seen_classifiers = []
        self._previous_rows = {}

    def swap_row(self, new_row):
        """ Stores new_row in the cache and returns the previous row
        with the same hash or None.
        """
        classifier = new_row.classifier
        with self._lock:
            self._pending.append(new_row)
            previous_row, self._previous_rows[classifier] = self._previous_rows.get(classifier, None), new_row
        return previous_row

    def flush(self, agg_pool):
        """ Generates a multi-insert statement and executes it against the analytics table. """

        # garbage collection routines
        self.garbage_collect_pending()
        self.garbage_collect_classifiers()

        # grab all the rows that need to be flushed
        with self._lock:
            flushing, self._pending = partition(self._pending, lambda r: r.valid())

        if len(flushing) > 0:
            self.record_classifiers(agg_pool, flushing)

            with agg_pool.connect() as conn:
                columns = ','.join(ANALYTICS_COLUMNS)
                while len(flushing):
                    batch, flushing = flushing[:128], flushing[128:]

                    row_values = sorted((row.analytics_values() for row in batch), key=lambda x: ('%s%s' % (x[0], x[1])))
                    query_params = sum(row_values, ())
                    values = ','.join(ANALYTICS_TEMPLATE * len(row_values))

                    conn.execute('''
                        INSERT INTO analytics (%s) VALUES %s
                        ON DUPLICATE KEY UPDATE value=value
                    ''' % (columns, values), *query_params)

    @throttle(60)
    def garbage_collect_pending(self):
        """ Check the pending array for any old rows, and then delete them """

        now = calendar.timegm(datetime.now().utctimetuple())
        with self._lock:
            self._pending = [r for r in self._pending if r.created > (now - 360)]

    def record_classifiers(self, agg_pool, analytic_rows):
        now = (util.sql_utcnow(),)  # 1-element tuple!
        value_template = ['(' + ','.join(['%s'] * len(CLASSIFIER_COLUMNS)) + ')']
        columns = ','.join(CLASSIFIER_COLUMNS)

        pending = []
        for row in analytic_rows:
            if row.classifier not in self._seen_classifiers:
                pending.append(row)
                self._seen_classifiers.append(row.classifier)

        if len(pending) > 0:
            with agg_pool.connect_master() as master_conn:
                while len(pending):
                    batch, pending = pending[:50], pending[50:]
                    row_values = [tuple(row.classifier_values()) + now for row in batch]
                    query_params = sum(row_values, ()) + now
                    values = ','.join(value_template * len(row_values))
                    sql = "INSERT INTO classifiers (%s) VALUES %s ON DUPLICATE KEY UPDATE last_updated=%%s" % (columns, values)
                    master_conn.execute(sql, *query_params)

    @throttle(60 * 5)
    def garbage_collect_classifiers(self):
        """
            The idea is to force the next record_classifiers to insert
            all of the classifiers that we are missing.  This ensures
            that we 'fix' any missing rows caused by garbage collection
            once every 5 minutes.
        """
        self._seen_classifiers = []


memsql_collectd/plugin.py
try:
    import collectd
except:
    pass

from memsql_collectd import __version__
from memsql.common.random_aggregator_pool import RandomAggregatorPool
from memsql.common.connection_pool import PoolConnectionException
from memsql_collectd import cluster, util
from memsql_collectd.analytics import AnalyticsCache, AnalyticsRow
from wraptor.decorators import throttle
import math
import re
import threading
import time
import os
import subprocess
import copy
import signal
import traceback

STATS_REPORT_INTERVAL = 5

# This is a data structure shared by every callback. We store all the
# configuration and globals in it.
class _AttrDict(dict):
    def __getattr__(self, name):
        try:
            return self[name]
        except KeyError:
            raise AttributeError(name)

    def __setattr__(self, name, val):
        self[name] = val

MEMSQL_DATA = _AttrDict(
    typesdb={},
    values_cache=AnalyticsCache(),
    node=None,
    counters=_AttrDict(stats=0, blacklisted=0, whitelisted=0, last_report=0),
    exiting=False
)

MEMSQL_DATA.config = _AttrDict(
    host=None,
    port=3306,
    user='root',
    password='',
    database='dashboard',
    memsqlnode=None,
    prefix=None,
    blacklist=set(),
    whitelist=set(),
    dfblacklist=True,
    derive=True,
    clusterhostname=None,
    skipdiskusage=False,
    skipcolumnardiskusage=False,
)

###################################
# Lifecycle functions

def memsql_config(config, data):
    """ Handle configuring collectd-memsql. """
    parsed_config = {
        'blacklist': set(),
        'whitelist': set()
    }
    for child in config.children:
        key = child.key.lower()
        val = child.values[0]
        if key in ('blacklist', 'whitelist'):
            parsed_config[key].add(val)
        else:
            if isinstance(val, (str, unicode)) and val.lower() in 'yes,true,on'.split(','):
                val = True
            elif isinstance(val, (str, unicode)) and val.lower() in 'no,false,off'.split(','):
                val = False
            parsed_config[key] = val

    for config_key in data.config:
        if config_key in ('blacklist', 'whitelist'):
            data.config[config_key] |= parsed_config[config_key]
        elif config_key in parsed_config:
            data.config[config_key] = parsed_config[config_key]

    if 'typesdb' in parsed_config:
        memsql_parse_types_file(parsed_config['typesdb'], data)

def memsql_init(data):
    """ Handles initializing collectd-memsql. """

    # handle SIGCHLD normally for subprocesses
    signal.signal(signal.SIGCHLD, signal.SIG_DFL)

    # verify we have required params
    assert len(data.typesdb) > 0, 'At least 1 TypesDB file path must be specified'
    assert data.config.host is not None, 'MemSQL host is not defined'
    assert data.config.port is not None, 'MemSQL port is not defined'
    if data.config.dfblacklist:
        data.config.blacklist |= set(['df.dev', 'df.dev-shm', 'df.run', 'df.run-lock', 'df.run-shm', 'df.proc'])
    for banned in data.config.blacklist:
        assert ' ' not in banned

    collectd.info('Initializing collectd-memsql %s with %s:%s' % (__version__, data.config.host, data.config.port))

    # initialize the aggregator pool
    data.pool = RandomAggregatorPool(
        host=data.config.host,
        port=int(data.config.port),
        user=data.config.user,
        password=data.config.password,
        database=data.config.database
    )

    data.config.blacklist = _compile_filter_re(data.config.blacklist)
    data.config.whitelist = _compile_filter_re(data.config.whitelist)

    try:
        # initialize a memsql node here if we can
        if data.config.memsqlnode or data.config.memsqlnode is None:
            throttled_find_node(data)
            if data.node is not None:
                data.node.connect().close()

        # initialize a pool connection
        data.pool.connect().close()
    except PoolConnectionException:
        pass

    # initialize the flushing thread
    data.flusher = FlushWorker(data)
    data.flusher.start()

    # initialize disk usage worker
    data.disk_crawler = DiskUsageWorker(data)
    data.disk_crawler.start()

def memsql_shutdown(data):
    """ Handles terminating collectd-memsql. """
    collectd.info('Terminating collectd-memsql')
    data.exiting = True

    collectd.info('Closing MemSQL connections...')
    data.pool.close()

    collectd.info('Terminating threads...')
    data.flusher.terminate()
    data.disk_crawler.terminate()

    collectd.info('Waiting for data flusher to exit...')
    data.flusher.join()
    collectd.info('Waiting for disk crawler to exit...')
    data.disk_crawler.join()

#########################
# Read/Write functions

def memsql_read(data):
    if data.exiting:
        return

    # update our node reference at least once every 60 seconds incase things change
    if data.config.memsqlnode or data.config.memsqlnode is None:
        throttled_find_node(data)

    if data.node is not None:
        memsql_status = collectd.Values(plugin="memsql", plugin_instance="status", type="gauge")
        for name, value in data.node.status():
            memsql_status.dispatch(type_instance=name, values=[value])

            if name in cluster.COUNTER_STATUS_VARIABLES:
                memsql_status.dispatch(type="counter", type_instance=name, values=[value])

        memsql_read_facts(data)
        memsql_read_disk_info(data, memsql_status)

@throttle(60)
def memsql_read_facts(data):
    if data.node.alias is not None:
        now = util.sql_utcnow()
        variables = list(data.node.variables())
        num_rows = len(variables)
        variables = sum([
            [now, data.node.alias, "memsql", "variables", gamma, value]
            for gamma, value in variables
        ], [])
        variables.append(now)

        if num_rows > 0:
            tmpl = ["(%s, %s, %s, %s, %s, %s)"]
            with data.pool.connect() as conn:
                sql = '''
                    INSERT INTO facts (last_updated, instance_id, alpha, beta, gamma, value) VALUES %s
                    ON DUPLICATE KEY UPDATE last_updated=%%s, value=VALUES(value)
                ''' % ",".join(tmpl * num_rows)
                conn.execute(sql, *variables)

@throttle(60)
def memsql_read_disk_info(data, memsql_status):
    memsql_dirs = data.node.directories()

    data_dir = memsql_dirs.get('Data_directory', '')
    dirs = {
        'data': data_dir,
        'segments': memsql_dirs.get('Segments_directory', os.path.join(data_dir, 'columns')),
        'logs': memsql_dirs.get('Logs_directory', os.path.join(data_dir, 'logs')),
        'snapshots': memsql_dirs.get('Snapshots_directory', os.path.join(data_dir, 'snapshots')),
        'plancache': memsql_dirs.get('Plancache_directory', os.path.join(data_dir, 'plancache'))
    }

    for label in dirs.keys():
        path = dirs[label]
        if not path or not os.path.exists(path) or not os.path.isdir(path):
            del dirs[label]

    if 'data' not in dirs or 'plancache' not in dirs:
        collectd.info("Can't find data or plancache directory.")
        return

    mounts = set()
    total_used = 0
    total_avail = 0
    for label, path in dirs.iteritems():
        paths = [path]
        try:
            # get columnar subdirs
            if label == 'segments':
                paths += os.listdir(path)

            with open(os.devnull, 'w') as devnull:
                p = subprocess.Popen(['df', '-k', '-P'] + paths, stderr=devnull, stdout=subprocess.PIPE)
                out = p.communicate()[0]

            for line in out.strip().split('\n')[1:]:
                fs, blocks, used, avail, percent, mount = line.strip().split(None, 5)
                if mount not in mounts:
                    mounts.add(mount)
                    total_used += int(used) * 1024
                    total_avail += int(avail) * 1024

        except (OSError, IndexError, ValueError) as e:
            collectd.info('Error in memsql_read_disk_info: ' + str(e))
            collectd.info(traceback.format_exc())

    memsql_status.dispatch(type_instance='used_storage_bytes', values=[total_used])
    memsql_status.dispatch(type_instance='available_storage_bytes', values=[total_avail])

    data.disk_crawler.update_dirs(dirs)
    for label, size in data.disk_crawler.disk_usage().iteritems():
        memsql_status.dispatch(type_instance=label + '_storage_bytes', values=[size])

def memsql_write(collectd_sample, data):
    """ Write handler for collectd.
    This function is called per sample taken from every plugin.  It is
    parallelized among multiple threads by collectd.
    """
    if data.exiting:
        return

    if data.node is not None:
        throttled_update_alias(data, collectd_sample)

    # get the value types for this sample
    types = data.typesdb
    if collectd_sample.type not in types:
        collectd.info('memsql_writer: do not know how to handle type %s. do you have all your types.db files configured?' % collectd_sample.type)
        return

    value_types = types[collectd_sample.type]

    if len(value_types) != len(collectd_sample.values):
        collectd.info('memsql_writer: differing number of values for type %s' % collectd_sample.type)
        return

    # for each value in this sample, insert it into the cache
    for (value_type, value) in zip(value_types, collectd_sample.values):
        cache_value(value, value_type[0], value_type[1], collectd_sample, data)

#########################
# Disk Usage thread

class DiskUsageWorker(threading.Thread):
    def __init__(self, data):
        super(DiskUsageWorker, self).__init__()
        self.data = data
        self._stop = threading.Event()
        self._disk_usage = {}
        self._disk_usage_lock = threading.RLock()

    def update_dirs(self, dirs):
        with self._disk_usage_lock:
            for label, path in dirs.iteritems():
                if label in self._disk_usage:
                    self._disk_usage[label]['path'] = path
                else:
                    self._disk_usage[label] = { 'path': path, 'bytes': None }
            for label in self._disk_usage:
                if label not in dirs:
                    del self._disk_usage[label]

    def disk_usage(self):
        out = {}
        with self._disk_usage_lock:
            for label, info in self._disk_usage.iteritems():
                if info['bytes'] is not None:
                    out[label] = info['bytes']
        return out

    def run(self):
        if self.data.config.skipdiskusage:
            return

        while not self._stop.isSet():
            time.sleep(5)

            if self.data.node is None:
                continue

            with self._disk_usage_lock:
                disk_usage = copy.deepcopy(self._disk_usage)

            for label, info in disk_usage.iteritems():
                if self.data.config.skipcolumnardiskusage and label == "segments":
                    continue

                try:
                    with open(os.devnull, 'w') as devnull:
                        p = subprocess.Popen(['du', '-k', '-s', info['path']], stderr=devnull, stdout=subprocess.PIPE)
                        out = p.communicate()[0]

                    size, _ = out.strip().split(None, 1)
                    size = int(size) * 1024

                    with self._disk_usage_lock:
                        if label in self._disk_usage:
                            self._disk_usage[label]['bytes'] = size
                except (OSError, IndexError, ValueError) as e:
                    collectd.info('Error in DiskUsageWorker: ' + str(e))
                    collectd.info(traceback.format_exc())
                    with self._disk_usage_lock:
                        if label in self._disk_usage:
                            self._disk_usage[label]['bytes'] = None

    def terminate(self):
        self._stop.set()

#########################
# Flushing thread

class FlushWorker(threading.Thread):
    def __init__(self, data):
        super(FlushWorker, self).__init__()
        self.data = data
        self._stop = threading.Event()

    def run(self):
        while not self._stop.isSet():
            start = time.time()
            try:
                self.data.values_cache.flush(self.data.pool)
            except Exception as e:
                collectd.error(str(e))

            # sleep up till the start of the next second
            sleep_time = math.floor(start + 1) - start
            time.sleep(sleep_time)

    def terminate(self):
        self._stop.set()

#########################
# Utility functions

@throttle(60)
def throttled_update_alias(data, collectd_sample):
    host = collectd_sample.host.replace('.', '_')
    data.node.update_alias(data.pool, host)

@throttle(60)
def throttled_find_node(data):
    node = cluster.find_node(data.pool, data.config.clusterhostname, data.config.memsqlnode)
    if data.node is not None and node is not None:
        # merge node details
        data.node.update_from_node(node)
    else:
        data.node = node

    if data.node is not None:
        collectd.info('I am a MemSQL node: %s:%s' % (data.node.host, data.node.port))

def memsql_parse_types_file(path, data):
    """ This function tries to parse a collectd compliant types.db file.
    Basically stolen from collectd-carbon.
    """
    types = data.typesdb

    f = open(path, 'r')

    for line in f:
        fields = line.split()
        if len(fields) < 2:
            continue

        type_name = fields[0]

        if type_name[0] == '#':
            continue

        v = []
        for ds in fields[1:]:
            ds = ds.rstrip(',')
            ds_fields = ds.split(':')

            if len(ds_fields) != 4:
                collectd.info('memsql_writer: cannot parse data source %s on type %s' % ( ds, type_name ))
                continue

            v.append(ds_fields)

        types[type_name] = v

    f.close()

def _compile_filter_re(filter_list):
    """Given a list of classifiers, like ['cpu.*', '*.r-kelly', '*carl*.*carl*'],
       return a regular expression which matches all strings which have prefixes in this
       list (where * is interpreted as [^\.]*?)
       In this case:
       \A(
         cpu\.[^\.]*?
        |[^\.]*?\.r\-kelly
        |[^\.]*?carl[^\.]*?\.[^\.]*?carl[^\.]*?
       )[\.\Z].*
    """
    if not filter_list:
        return None
    return re.compile(r'\A(' +
        '|'.join(re.escape(item).replace(r'\*', r'[^\.]*?') for item in filter_list)
        + r')(\.|\Z)')

def _blacklisted(data, classifier_str):
    return data.config.blacklist and data.config.blacklist.match(classifier_str)

def _whitelisted(data, classifier_str):
    return (not data.config.whitelist) or data.config.whitelist.match(classifier_str)


def cache_value(new_value, data_source_name, data_source_type, collectd_sample, data):
    """ Cache a new value into the analytics cache.  Handles converting
    COUNTER and DERIVE types.
    """

    data.counters.stats += 1
    now = time.time()
    if (now - data.counters.last_report > STATS_REPORT_INTERVAL):
        data.counters.last_report = now
        collectd.info("%d stats collected, %d excluded by whitelist, %d blacklisted" % (
            data.counters.stats,
            data.counters.whitelisted,
            data.counters.blacklisted
        ))

    temp_classifier = [
        data.config.prefix,         # alpha... (None's will be filtered)
        collectd_sample.plugin,
        collectd_sample.plugin_instance,
        collectd_sample.type,
        collectd_sample.type_instance,
        data_source_name,
    ]

    # we do this since '.' could be inside one of the classifier parts (like plugin, plugin_instance, etc)
    classifier_str = '.'.join(filter(None, temp_classifier))
    classifier = classifier_str.split('.')

    if not _whitelisted(data, classifier_str):
        data.counters.whitelisted += 1
        return

    if _blacklisted(data, classifier_str):
        data.counters.blacklisted += 1
        return

    instance_id = collectd_sample.host
    if instance_id is not None:
        instance_id = instance_id.replace('.', '_')

    new_row = AnalyticsRow(
        int(collectd_sample.time),
        new_value,
        instance_id,
        *classifier
    )

    previous_row = data.values_cache.swap_row(new_row)

    if data_source_type not in ['ABSOLUTE', 'COUNTER', 'DERIVE', 'GAUGE']:
        collectd.info("MemSQL: Undefined data source %s" % data_source_type)
        return

    # special case counter and derive since we only care about the
    # *difference* between their values (rather than their actual value)
    if data_source_type in ['ABSOLUTE', 'GAUGE'] or not data.config.derive:
        new_row.value = new_value
    elif data_source_type in ['COUNTER', 'DERIVE']:
        if previous_row is None:
            return

        old_value = previous_row.raw_value

        # don't let the time delta fall below 1
        time_delta = float(max(1, new_row.created - previous_row.created))

        # The following formula handles wrapping COUNTER data types
        # around since COUNTERs should never be negative.
        # Taken from: https://collectd.org/wiki/index.php/Data_source
        if data_source_type == 'COUNTER' and new_value < old_value:
            if old_value < 2 ** 32:
                new_row.value = (2 ** 32 - old_value + new_value) / time_delta
            else:
                new_row.value = (2 ** 64 - old_value + new_value) / time_delta

            if collectd_sample.plugin == 'cpu' and collectd_sample.type == 'cpu' and collectd_sample.type_instance == 'wait':
                # in virtualized environments, iowait sometimes wraps around and then back
                if new_row.value > (2 ** 31):
                    new_row.raw_value = previous_row.raw_value
                    new_row.value = previous_row.value
        else:
            # the default wrap-around formula
            new_row.value = (new_value - old_value) / time_delta

######################
# Register callbacks

try:
    collectd.register_config(memsql_config, MEMSQL_DATA)
    collectd.register_init(memsql_init, MEMSQL_DATA)
    collectd.register_write(memsql_write, MEMSQL_DATA)
    collectd.register_shutdown(memsql_shutdown, MEMSQL_DATA)
    collectd.register_read(memsql_read, 1, MEMSQL_DATA)
except:
    # collectd not available
    pass
